base/primes.jl

```diff
@@ -73,51 +74,72 @@ isprime(n::UInt128) =
 isprime(n::Int128) = n < 2 ? false :
     n <= typemax(Int64)  ? isprime(Int64(n))  : isprime(BigInt(n))
 
-# TODO: faster factorization algorithms?
-
-const PRIMES = primes(10000)
 
+# Trial division of small (< 2^16) precomputed primes +
+# Pollard rho's algorithm with Richard P. Brent optimizations
+#     http://en.wikipedia.org/wiki/Trial_division
+#     http://en.wikipedia.org/wiki/Pollard%27s_rho_algorithm
+#     http://maths-people.anu.edu.au/~brent/pub/pub051.html
+#
 function factor{T<:Integer}(n::T)
     0 < n || throw(ArgumentError("number to be factored must be ≥ 0, got $n"))
     h = Dict{T,Int}()
     n == 1 && return h
-    n <= 3 && (h[n] = 1; return h)
-    local s::T, p::T
-    s = isqrt(n)
+    isprime(n) && (h[n] = 1; return h)
+    local p::T
     for p in PRIMES
-        if p > s
-            h[n] = 1
-            return h
-        end
         if n % p == 0
+            h[p] = get(h,p,0)+1
+            n = oftype(n, div(n,p))
             while n % p == 0
                 h[p] = get(h,p,0)+1
                 n = oftype(n, div(n,p))
             end
             n == 1 && return h
-            if isprime(n)
-                h[n] = 1
-                return h
-            end
-            s = isqrt(n)
+            isprime(n) && (h[n] = 1; return h)
         end
     end
-    p = PRIMES[end]+2
-    while p <= s
-        if n % p == 0
-            while n % p == 0
-                h[p] = get(h,p,0)+1
-                n = oftype(n, div(n,p))
+    pollardfactors!(n, h)
+end
+
+function pollardfactors!{T<:Integer}(n::T, h::Dict{T,Int})
+    while true
+        local c::T = rand(1:(n-1)), G::T = 1, r::T = 1, y::T = rand(0:(n-1)), m::T = 1900
+        local ys::T, q::T = 1, x::T
+        while c == n - 2
+            c = rand(1:(n-1))
+        end
+        while G == 1
+            x = y
+            for i in 1:r
+                y = oftype(y, widemul(y,y)%n)
```

Comment:
The `oftype` is only necessary for the cases where you are assigning to one of the arguments, since `function foo(x::T) .... ` allows you to change the type of `x` in the body of `foo`.